compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1654,6 +1654,102 @@ fn add_post_link_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor
+/// participate in linking like object files, but this proves to be expensive (#93791). Therefore
+/// we instead just introduce an undefined reference to them. This could be done by `-u` command
+/// line option to the linker or `EXTERN(...)` in linker scripts, however they does not only
+/// introduce an undefined reference, but also make them the GC roots, preventing `--gc-sections`
+/// from removing them, and this is especially problematic for embedded programming where every
+/// byte counts.
+///
+/// This method creates a synthetic object file, which contains undefined references to all symbols
+/// that are necessary for the linking. They are all placed in a section that we both 1) mark as
+/// debug and 2) mark as exclude. The linker will therefore pick relevant rlibs for linking, but
+/// this section will be discarded later, and unused `#[no_mangle]` or `#[used]` can still be
+/// discard by GC sections.
+fn add_linked_symbol_object(
+    cmd: &mut dyn Linker,
+    sess: &Session,
+    tmpdir: &Path,
+    symbols: &[(String, bool)],
+) {
+    use object::{BinaryFormat, SectionFlags};
+
+    if symbols.is_empty() {
+        return;
+    }
+
+    let Some(mut file) = super::metadata::create_object_file(sess) else {
+        return;
+    };
+
+    let section = file.add_section(
+        file.segment_name(object::write::StandardSegment::Debug).to_vec(),
+        b".rsym".to_vec(),
+        object::SectionKind::Debug,
+    );
+    match file.format() {
+        BinaryFormat::Coff => {
+            file.section_mut(section).flags =
+                SectionFlags::Coff { characteristics: object::pe::IMAGE_SCN_LNK_REMOVE };
+        }
+        BinaryFormat::Elf => {
+            file.section_mut(section).flags =
+                SectionFlags::Elf { sh_flags: object::elf::SHF_EXCLUDE as u64 };
+        }
+        _ => {}
+    };
+
+    // Create enough space for relocation of each symbol.
+    file.append_section_data(
+        section,
+        &vec![0; symbols.len() * sess.target.pointer_width as usize],
+        sess.target.pointer_width as u64 / 8,
+    );
+
+    for (i, (sym, tls)) in symbols.iter().enumerate() {
+        let sym_id = file.add_symbol(object::write::Symbol {
+            name: sym.clone().into(),
+            value: 0,
+            size: 0,
+            kind: if *tls { object::SymbolKind::Tls } else { object::SymbolKind::Unknown },
+            scope: object::SymbolScope::Unknown,
+            weak: false,
+            section: object::write::SymbolSection::Undefined,
+            flags: object::SymbolFlags::None,
+        });
+        file.add_relocation(
+            section,
+            object::write::Relocation {
+                offset: (i as u64) * sess.target.pointer_width as u64 / 8,
+                size: sess.target.pointer_width as u8,
+                kind: object::RelocationKind::Absolute,
+                encoding: object::RelocationEncoding::Generic,
```

Comment:
This is technically not valid for TLS symbols. They need a specific platform dependent relocation kind AFAIK. Some linkers may allow this in discarded sections, but I'm not sure all of them do.

Reply:
I just realized that the relocation may actually be superficial -- at least for ELF targets:
* The ELF spec reads:
  > When the link editor searches archive libraries, it extracts archive members that contain 
definitions of undefined global symbols.
  
  there is no mention that it needs to be used symbol.
* I just verified that both GNU ld and LLD behaves the same when I remove the relocation part and just keep the symbol table.

I couldn't verify the behaviour of Mach-O and COFF.

Reply:
I mean the relocation itself is not valid for TLS symbols AFAIK.

Reply:
Yeah, so removing the relocation at all would avoid the issue.

Reply:
Plus I think the relocation is valid for a few platforms, but I understand that it could be invalid for many others.

Reply:
Verified COFF with MSVC linker, same behaviour as expected. I don't have any Mac device so couldn't verify Mach-O behaviour, but presumably it'll be the same.